packages/vite/types/hmrPayload.d.ts

```diff
@@ -20,6 +20,7 @@ export interface Update {
   path: string
   acceptedPath: string
   timestamp: number
+  explicitImportRequired: boolean | undefined
```

Comment:
Thanks Sapphi, amazing how long this bug was open. About
> I passed explicitImportRequired property to do the check on server side, not to duplicate isExplicitImportRequired function to client/client.ts.

I think that given `hmrPayload` is part of the public API, and that `isExplicitImportRequired` isn't a complex function, it is better, in this case, to duplicate it in the client.

Reply:
Done 👍 
In Vite 4, maybe we could change this like what I did to reduce maintenance burden.

Reply:
Given that `explicitImportRequired` can be derived from `path`, I think that it is better to avoid it. Maybe we could find a better way to share code between the client and node? Maybe we need a `shared.ts` utils file that only has code that can be run in both?

Reply:
I missed in my previous comment that `isJSRequest` was depending on the hardcoded list for `knownJSImports`. Now I better understand why you decided to add the field. 
Maybe we could add the shared util file, now? cc @bluwy interested in your opinion here too

Reply:
FWIW perhaps we want to stick to sapphi's initial implementation if we still want to consider https://github.com/vitejs/vite/pull/3828 (allow specify specific extensions to be considered js which doesn't require the `?import`). And this can only be configured on node side.

Inlining the code inside client or with shared.ts would prevent future configuration if needed.

Reply:
Yes, sorry I make you do extra work here. @bluwy has a good point. I'm thinking if we should call only pass `knownJSSrc`, but I'm also leaning now towards your original implementation. I'm not fond of the `explicitImportRequired` name, but I can't find a better one either. I think we should revert the last commit, and we can add this to discuss it in the next meeting.

Reply:
Never mind🙂